tests/lib/fortigate_test.py

```diff
@@ -0,0 +1,352 @@
+                                    src_port='HTTP',
+                                    remove_fields=('dest_port',))
+    icmp_term = self.fmt.format(TERM_TEMPLATE,
+                                protocol='icmp',
+                                remove_fields=('dest_port', 'src_port'))
+    ip_term = self.fmt.format(TERM_TEMPLATE,
+                              remove_fields=('dest_port',
+                                             'src_port', 'protocol'))
+    custom_port_term = self.fmt.format(TERM_TEMPLATE, src_port='WHOIS')
+
+    dup_acl = fortigate.Fortigate(policy.ParsePolicy(
+        GOOD_HEADER + dup_port_term,
+        self.naming), EXP_INFO)
+    diff_acl = fortigate.Fortigate(policy.ParsePolicy(
+        GOOD_HEADER + diff_port_term,
+        self.naming), EXP_INFO)
+    src_only_acl = fortigate.Fortigate(policy.ParsePolicy(
+        GOOD_HEADER + src_only_term,
+        self.naming), EXP_INFO)
+    icmp_acl = fortigate.Fortigate(policy.ParsePolicy(
+        GOOD_HEADER + icmp_term,
+        self.naming), EXP_INFO)
+    ip_acl = fortigate.Fortigate(policy.ParsePolicy(
+        GOOD_HEADER + ip_term,
+        self.naming), EXP_INFO)
+    custom_port_acl = fortigate.Fortigate(
+        policy.ParsePolicy(GOOD_HEADER + custom_port_term,
+                           self.naming), EXP_INFO)
+
+    dup_sig = 'set service HTTP\n'
+    diff_sig = 'set service HTTP HTTPS\n'
+    src_only_sig = dup_sig
+    icmp_sig = 'set service ALL_ICMP\n'
+    ip_sig = 'set service ALL\n'
+    custom_port_sig = ('config firewall service custom\n\tedit 43\n\t\t' 
+                       'set protocol TCP/UDP\n\t\tset tcp-portrange 43\n\tnext')
+
+    self.assertIn(
+        dup_sig, str(dup_acl), '[%s]' % str(dup_acl))
+    self.assertIn(
+        diff_sig, str(diff_acl), '[%s]' % str(diff_acl))
+    self.assertIn(
+        src_only_sig, str(src_only_acl), '[%s]' % str(src_only_acl))
+    self.assertIn(
+        icmp_sig, str(icmp_acl), '[%s]' % str(icmp_acl))
+    self.assertIn(
+        ip_sig , str(ip_acl), '[%s]' % str(ip_acl))
+    self.assertIn(
+        custom_port_sig, str(custom_port_acl), '[%s]' % str(custom_port_acl))
+
+  def testInterfaces(self):
+    no_interfaces_term = self.fmt.format(TERM_TEMPLATE,
+                                         remove_fields=('src_interface',
+                                                        'dest_interface'))
+    src_only_int_term = self.fmt.format(TERM_TEMPLATE,
+                                        src_interface='wan1',
+                                        remove_fields=('dest_interface',))
+    dest_only_int_term = self.fmt.format(TERM_TEMPLATE,
+                                         dest_interface='wan2',
+                                         remove_fields=('src_interface',))
+    both_interfaces_term = self.fmt.format(TERM_TEMPLATE,
+                                           src_interface='wan1',
+                                           dest_interface='wan2', )
```

Comment:
Not sure why the end comma's here but at least on this line please remove the end space.